// @electron-forge/core/dist/api/init-scripts/init-npm.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.siblingDep = siblingDep;
exports.initNPM = exports.exactDevDeps = exports.devDeps = exports.deps = void 0;
var _path = _interopRequireDefault(require("path"));
var _coreUtils = require("@electron-forge/core-utils");
var _debug = _interopRequireDefault(require("debug"));
var _fsExtra = _interopRequireDefault(require("fs-extra"));
var _installDependencies = _interopRequireWildcard(require("../../util/install-dependencies"));
var _readPackageJson = require("../../util/read-package-json");
function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _interopRequireWildcard(obj) {
    if (obj && obj.__esModule) {
        return obj;
    } else {
        var newObj = {};
        if (obj != null) {
            for(var key in obj){
                if (Object.prototype.hasOwnProperty.call(obj, key)) {
                    var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {};
                    if (desc.get || desc.set) {
                        Object.defineProperty(newObj, key, desc);
                    } else {
                        newObj[key] = obj[key];
                    }
                }
            }
        }
        newObj.default = obj;
        return newObj;
    }
}
const d = (0, _debug).default('electron-forge:init:npm');
const corePackage = _fsExtra.default.readJsonSync(_path.default.resolve(__dirname, '../../../package.json'));
function siblingDep(name) {
    return `@electron-forge/${name}@^${corePackage.version}`;
}
const deps = [
    'electron-squirrel-startup'
];
exports.deps = deps;
const devDeps = [
    siblingDep('cli'),
    siblingDep('maker-squirrel'),
    siblingDep('maker-zip'),
    siblingDep('maker-deb'),
    siblingDep('maker-rpm')
];
exports.devDeps = devDeps;
const exactDevDeps = [
    'electron'
];
exports.exactDevDeps = exactDevDeps;
const initNPM = async (dir, task)=>{
    d('installing dependencies');
    const packageManager = (0, _coreUtils).safeYarnOrNpm();
    task.output = `${packageManager} install ${deps.join(' ')}`;
    await (0, _installDependencies).default(dir, deps);
    d('installing devDependencies');
    task.output = `${packageManager} install --dev ${deps.join(' ')}`;
    await (0, _installDependencies).default(dir, devDeps, _installDependencies.DepType.DEV);
    d('installing exact devDependencies');
    for (const packageName of exactDevDeps){
        task.output = `${packageManager} install --dev --exact ${packageName}`;
        await (0, _installDependencies).default(dir, [
            packageName
        ], _installDependencies.DepType.DEV, _installDependencies.DepVersionRestriction.EXACT);
    }
    // This logic allows developers working on forge itself to easily init
    // a local template and have it use their local plugins / core / cli packages
    if (process.env.LINK_FORGE_DEPENDENCIES_ON_INIT) {
        const packageJson = await (0, _readPackageJson).readRawPackageJson(dir);
        const linkFolder = _path.default.resolve(__dirname, '..', '..', '..', '..', '..', '..', '.links');
        for (const packageName of Object.keys(packageJson.devDependencies)){
            if (packageName.startsWith('@electron-forge/')) {
                task.output = `${packageManager} link --link-folder ${linkFolder} ${packageName}`;
                await (0, _coreUtils).yarnOrNpmSpawn([
                    'link',
                    '--link-folder',
                    linkFolder,
                    packageName
                ], {
                    cwd: dir
                });
            }
        }
    }
};
exports.initNPM = initNPM;
