// @electron-forge/core/dist/api/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "ElectronProcess", {
    enumerable: true,
    get: function() {
        return _sharedTypes.ElectronProcess;
    }
});
Object.defineProperty(exports, "ForgeMakeResult", {
    enumerable: true,
    get: function() {
        return _sharedTypes.ForgeMakeResult;
    }
});
Object.defineProperty(exports, "ForgeUtils", {
    enumerable: true,
    get: function() {
        return _util.default;
    }
});
Object.defineProperty(exports, "ImportOptions", {
    enumerable: true,
    get: function() {
        return _import.ImportOptions;
    }
});
Object.defineProperty(exports, "InitOptions", {
    enumerable: true,
    get: function() {
        return _init.InitOptions;
    }
});
Object.defineProperty(exports, "MakeOptions", {
    enumerable: true,
    get: function() {
        return _make.MakeOptions;
    }
});
Object.defineProperty(exports, "PackageOptions", {
    enumerable: true,
    get: function() {
        return _package.PackageOptions;
    }
});
Object.defineProperty(exports, "PublishOptions", {
    enumerable: true,
    get: function() {
        return _publish.PublishOptions;
    }
});
Object.defineProperty(exports, "StartOptions", {
    enumerable: true,
    get: function() {
        return _start.StartOptions;
    }
});
exports.utils = exports.api = void 0;
var _sharedTypes = require("@electron-forge/shared-types");
var _util = _interopRequireDefault(require("../util"));
var _import = _interopRequireWildcard(require("./import"));
var _init = _interopRequireWildcard(require("./init"));
var _make = _interopRequireWildcard(require("./make"));
var _package = _interopRequireWildcard(require("./package"));
var _publish = _interopRequireWildcard(require("./publish"));
var _start = _interopRequireWildcard(require("./start"));
function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _interopRequireWildcard(obj) {
    if (obj && obj.__esModule) {
        return obj;
    } else {
        var newObj = {};
        if (obj != null) {
            for(var key in obj){
                if (Object.prototype.hasOwnProperty.call(obj, key)) {
                    var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {};
                    if (desc.get || desc.set) {
                        Object.defineProperty(newObj, key, desc);
                    } else {
                        newObj[key] = obj[key];
                    }
                }
            }
        }
        newObj.default = obj;
        return newObj;
    }
}
class ForgeAPI {
    /**
   * Attempt to import a given module directory to the Electron Forge standard.
   *
   * * Sets up `git` and the correct NPM dependencies
   * * Adds a template forge config to `package.json`
   */ import(opts) {
        return (0, _import).default(opts);
    }
    /**
   * Initialize a new Electron Forge template project in the given directory.
   */ init(opts) {
        return (0, _init).default(opts);
    }
    /**
   * Make distributables for an Electron application
   */ make(opts) {
        return (0, _make).default(opts);
    }
    /**
   * Resolves hooks if they are a path to a file (instead of a `Function`)
   */ async package(opts) {
        await (0, _package).default(opts);
    }
    /**
   * Publish an Electron application into the given target service
   */ publish(opts) {
        return (0, _publish).default(opts);
    }
    /**
   * Start an Electron application.
   *
   * Handles things like native module rebuilding for you on the fly
   */ start(opts) {
        return (0, _start).default(opts);
    }
}
exports.ForgeAPI = ForgeAPI;
const api = new ForgeAPI();
exports.api = api;
const utils = new _util.default();
exports.utils = utils;
