// @electron-forge/maker-squirrel/dist/MakerSquirrel.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.MakerSquirrel = exports.default = void 0;
var _path = _interopRequireDefault(require("path"));
var _makerBase = require("@electron-forge/maker-base");
var _electronWinstaller = require("electron-winstaller");
var _fsExtra = _interopRequireDefault(require("fs-extra"));
function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
class MakerSquirrel extends _makerBase.MakerBase {
    isSupportedOnCurrentPlatform() {
        return this.isInstalled('electron-winstaller') && !process.env.DISABLE_SQUIRREL_TEST;
    }
    async make({ dir , makeDir , targetArch , packageJSON , appName , forgeConfig  }) {
        const outPath = _path.default.resolve(makeDir, `squirrel.windows/${targetArch}`);
        await this.ensureDirectory(outPath);
        const winstallerConfig = {
            name: typeof packageJSON.name === 'string' ? packageJSON.name.replace(/-/g, '_') : undefined,
            title: appName,
            noMsi: true,
            exe: `${forgeConfig.packagerConfig.executableName || appName}.exe`,
            setupExe: `${appName}-${packageJSON.version} Setup.exe`,
            ...this.config,
            appDirectory: dir,
            outputDirectory: outPath
        };
        await (0, _electronWinstaller).createWindowsInstaller(winstallerConfig);
        const nupkgVersion = (0, _electronWinstaller).convertVersion(packageJSON.version);
        const artifacts = [
            _path.default.resolve(outPath, 'RELEASES'),
            _path.default.resolve(outPath, winstallerConfig.setupExe || `${appName}Setup.exe`),
            _path.default.resolve(outPath, `${winstallerConfig.name}-${nupkgVersion}-full.nupkg`), 
        ];
        const deltaPath = _path.default.resolve(outPath, `${winstallerConfig.name}-${nupkgVersion}-delta.nupkg`);
        if (winstallerConfig.remoteReleases || await _fsExtra.default.pathExists(deltaPath)) {
            artifacts.push(deltaPath);
        }
        const msiPath = _path.default.resolve(outPath, winstallerConfig.setupMsi || `${appName}Setup.msi`);
        if (!winstallerConfig.noMsi && await _fsExtra.default.pathExists(msiPath)) {
            artifacts.push(msiPath);
        }
        return artifacts;
    }
    constructor(...args){
        super(...args);
        this.name = 'squirrel';
        this.defaultPlatforms = [
            'win32'
        ];
    }
}
exports.default = MakerSquirrel;
exports.MakerSquirrel = MakerSquirrel;
