// @electron-forge/template-webpack-typescript/dist/WebpackTypeScriptTemplate.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = void 0;
var _path = _interopRequireDefault(require("path"));
var _templateBase = require("@electron-forge/template-base");
var _fsExtra = _interopRequireDefault(require("fs-extra"));
function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
class WebpackTypeScriptTemplate extends _templateBase.BaseTemplate {
    async initializeTemplate(directory, options) {
        const superTasks = await super.initializeTemplate(directory, options);
        return [
            ...superTasks,
            {
                title: 'Setting up Forge configuration',
                task: async ()=>{
                    await this.copyTemplateFile(directory, 'forge.config.ts');
                    await _fsExtra.default.remove(_path.default.resolve(directory, 'forge.config.js'));
                }
            },
            {
                title: 'Preparing TypeScript files and configuration',
                task: async ()=>{
                    const filePath = (fileName)=>_path.default.join(directory, 'src', fileName)
                    ;
                    // Copy Webpack files
                    await this.copyTemplateFile(directory, 'webpack.main.config.ts');
                    await this.copyTemplateFile(directory, 'webpack.renderer.config.ts');
                    await this.copyTemplateFile(directory, 'webpack.rules.ts');
                    await this.copyTemplateFile(directory, 'webpack.plugins.ts');
                    await this.updateFileByLine(_path.default.resolve(directory, 'src', 'index.html'), (line)=>{
                        if (line.includes('link rel="stylesheet"')) return '';
                        return line;
                    });
                    // Copy tsconfig with a small set of presets
                    await this.copyTemplateFile(directory, 'tsconfig.json');
                    // Copy eslint config with recommended settings
                    await this.copyTemplateFile(directory, '.eslintrc.json');
                    // Remove index.js and replace with index.ts
                    await _fsExtra.default.remove(filePath('index.js'));
                    await this.copyTemplateFile(_path.default.join(directory, 'src'), 'index.ts');
                    await this.copyTemplateFile(_path.default.join(directory, 'src'), 'renderer.ts');
                    // Remove preload.js and replace with preload.ts
                    await _fsExtra.default.remove(filePath('preload.js'));
                    await this.copyTemplateFile(_path.default.join(directory, 'src'), 'preload.ts');
                    // update package.json
                    const packageJSONPath = _path.default.resolve(directory, 'package.json');
                    const packageJSON = await _fsExtra.default.readJson(packageJSONPath);
                    packageJSON.main = '.webpack/main';
                    // Configure scripts for TS template
                    packageJSON.scripts.lint = 'eslint --ext .ts,.tsx .';
                    await _fsExtra.default.writeJson(packageJSONPath, packageJSON, {
                        spaces: 2
                    });
                }
            }, 
        ];
    }
    constructor(...args){
        super(...args);
        this.templateDir = _path.default.resolve(__dirname, '..', 'tmpl');
    }
}
var _default = new WebpackTypeScriptTemplate();
exports.default = _default;
