// @electron-forge/publisher-base/dist/Publisher.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.PublisherBase = exports.default = void 0;
class Publisher {
    /**
   * @param config - A configuration object for this publisher
   * @param platformsToPublishOn - If you want this maker to run on platforms different from `defaultPlatforms` you can provide those platforms here
   */ constructor(config, platformsToPublishOn){
        this.config = config;
        this.platformsToPublishOn = platformsToPublishOn;
        this.config = config;
        Object.defineProperty(this, '__isElectronForgePublisher', {
            value: true,
            enumerable: false,
            configurable: false
        });
    }
    get platforms() {
        if (this.platformsToPublishOn) return this.platformsToPublishOn;
        if (this.defaultPlatforms) return this.defaultPlatforms;
        return [
            'win32',
            'linux',
            'darwin',
            'mas'
        ];
    }
    /**
   * Publishers must implement this method to publish the artifacts returned from
   * make calls.  If any errors occur you must throw them, failing silently or simply
   * logging will not propagate issues up to forge.
   *
   * Please note for a given version publish will be called multiple times, once
   * for each set of "platform" and "arch".  This means if you are publishing
   * darwin and win32 artifacts to somewhere like GitHub on the first publish call
   * you will have to create the version on GitHub and the second call will just
   * be appending files to the existing version.
   */ // eslint-disable-next-line @typescript-eslint/no-unused-vars
    async publish(opts) {
        throw new Error(`Publisher ${this.name} did not implement the publish method`);
    }
}
exports.default = Publisher;
exports.PublisherBase = Publisher;
