// @electron-forge/core/dist/util/install-dependencies.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = exports.DepVersionRestriction = exports.DepType = void 0;
var _coreUtils = require("@electron-forge/core-utils");
var _crossSpawnPromise = require("@malept/cross-spawn-promise");
var _debug = _interopRequireDefault(require("debug"));
function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const d = (0, _debug).default('electron-forge:dependency-installer');
var DepType;
exports.DepType = DepType;
(function(DepType) {
    DepType["PROD"] = "PROD";
    DepType["DEV"] = "DEV";
})(DepType || (exports.DepType = DepType = {}));
var DepVersionRestriction;
exports.DepVersionRestriction = DepVersionRestriction;
(function(DepVersionRestriction) {
    DepVersionRestriction["EXACT"] = "EXACT";
    DepVersionRestriction["RANGE"] = "RANGE";
})(DepVersionRestriction || (exports.DepVersionRestriction = DepVersionRestriction = {}));
var _default = async (dir, deps, depType = DepType.PROD, versionRestriction = DepVersionRestriction.RANGE)=>{
    d('installing', JSON.stringify(deps), 'in:', dir, `depType=${depType},versionRestriction=${versionRestriction},withYarn=${(0, _coreUtils).hasYarn()}`);
    if (deps.length === 0) {
        d('nothing to install, stopping immediately');
        return Promise.resolve();
    }
    let cmd = [
        'install'
    ].concat(deps);
    if ((0, _coreUtils).hasYarn()) {
        cmd = [
            'add'
        ].concat(deps);
        if (depType === DepType.DEV) cmd.push('--dev');
        if (versionRestriction === DepVersionRestriction.EXACT) cmd.push('--exact');
    } else {
        if (versionRestriction === DepVersionRestriction.EXACT) cmd.push('--save-exact');
        if (depType === DepType.DEV) cmd.push('--save-dev');
        if (depType === DepType.PROD) cmd.push('--save');
    }
    d('executing', JSON.stringify(cmd), 'in:', dir);
    try {
        await (0, _coreUtils).yarnOrNpmSpawn(cmd, {
            cwd: dir,
            stdio: 'pipe'
        });
    } catch (err) {
        if (err instanceof _crossSpawnPromise.ExitError) {
            throw new Error(`Failed to install modules: ${JSON.stringify(deps)}\n\nWith output: ${err.message}\n${err.stderr ? err.stderr.toString() : ''}`);
        } else {
            throw err;
        }
    }
};
exports.default = _default;
