// @electron-forge/template-webpack/dist/WebpackTemplate.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = void 0;
var _path = _interopRequireDefault(require("path"));
var _templateBase = require("@electron-forge/template-base");
var _fsExtra = _interopRequireDefault(require("fs-extra"));
function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
class WebpackTemplate extends _templateBase.BaseTemplate {
    async initializeTemplate(directory, options) {
        const superTasks = await super.initializeTemplate(directory, options);
        return [
            ...superTasks,
            {
                title: 'Setting up Forge configuration',
                task: async ()=>{
                    await this.copyTemplateFile(directory, 'forge.config.js');
                }
            },
            {
                title: 'Setting up webpack configuration',
                task: async ()=>{
                    await this.copyTemplateFile(directory, 'webpack.main.config.js');
                    await this.copyTemplateFile(directory, 'webpack.renderer.config.js');
                    await this.copyTemplateFile(directory, 'webpack.rules.js');
                    await this.copyTemplateFile(_path.default.join(directory, 'src'), 'renderer.js');
                    await this.copyTemplateFile(_path.default.join(directory, 'src'), 'preload.js');
                    await this.updateFileByLine(_path.default.resolve(directory, 'src', 'index.js'), (line)=>{
                        if (line.includes('mainWindow.loadFile')) return '  mainWindow.loadURL(MAIN_WINDOW_WEBPACK_ENTRY);';
                        if (line.includes('preload: ')) return '      preload: MAIN_WINDOW_PRELOAD_WEBPACK_ENTRY,';
                        return line;
                    }, _path.default.resolve(directory, 'src', 'main.js'));
                    await this.updateFileByLine(_path.default.resolve(directory, 'src', 'index.html'), (line)=>{
                        if (line.includes('link rel="stylesheet"')) return '';
                        return line;
                    });
                    // update package.json entry point
                    const pjPath = _path.default.resolve(directory, 'package.json');
                    const currentPJ = await _fsExtra.default.readJson(pjPath);
                    currentPJ.main = '.webpack/main';
                    await _fsExtra.default.writeJson(pjPath, currentPJ, {
                        spaces: 2
                    });
                }
            }, 
        ];
    }
    constructor(...args){
        super(...args);
        this.templateDir = _path.default.resolve(__dirname, '..', 'tmpl');
    }
}
var _default = new WebpackTemplate();
exports.default = _default;
