// @electron-forge/plugin-base/dist/Plugin.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "StartOptions", {
    enumerable: true,
    get: function() {
        return _sharedTypes.StartOptions;
    }
});
exports.namedHookWithTaskFn = exports.PluginBase = exports.default = void 0;
var _sharedTypes = require("@electron-forge/shared-types");
class Plugin {
    constructor(config){
        this.config = config;
        /** @internal */ this._resolvedHooks = {};
        Object.defineProperty(this, '__isElectronForgePlugin', {
            value: true,
            enumerable: false,
            configurable: false
        });
    }
    init(_dir, _config) {
        // This logic ensures that we only call getHooks once regardless of how many
        // times we trip hook logic in the PluginInterface.
        this._resolvedHooks = this.getHooks();
        this.getHooks = ()=>this._resolvedHooks
        ;
    }
    getHooks() {
        return {};
    }
    async startLogic(_startOpts) {
        return false;
    }
}
exports.default = Plugin;
exports.PluginBase = Plugin;
const namedHookWithTaskFn = (hookFn, name)=>{
    function namedHookWithTaskInner(...args) {
        return hookFn(this, ...args);
    }
    const fn = namedHookWithTaskInner;
    fn.__hookName = name;
    return fn;
};
exports.namedHookWithTaskFn = namedHookWithTaskFn;
